### PR TITLE
Load .glyphs instances and use them to inform axis rigging

### PR DESCRIPTION
**Pull request by rsheeter**
Improve, though not yet to parity with fontmake, handling of glyphs 2/3 instances:

* Load glyphs instances for glyphs v2 and v3
   * Upgrade v2 to v3 per our std practice
   * We do not yet support all fields. Baby steps.
   * The code for this is tiresome and lengthy
* Fix typo that lead to avar not being merged into final font
* Include instance mappings in user:design mappings, which improves avar production
   * We now match the Oswald avar noted in https://github.com/googlefonts/fontmake-rs/issues/250#issuecomment-1504057038
* Figure out the default instance and use it to assign axis defaults
   * We now match the Oswald fvar axes noted in https://github.com/googlefonts/fontmake-rs/issues/250#issuecomment-1504057038
   * We do not yet emit any named instances

Step toward #250

### CODE REVIEW

File: fontc/src/lib.rs [hunk 1]
```diff
@@ -1422,4 +1424,81 @@ mod tests {
...
+            .map(|m| m.unwrap())
+            .map(|m| {
+                m.axis_value_maps()
+                    .iter()
+                    .map(|e| (e.from_coordinate().to_f32(), e.to_coordinate().to_f32()))
+                    .collect::<Vec<_>>()
+            })
+            .collect();
+
+        // 0 is 400, 7 is 700, .33 => .66 is mapping 500 => 600 as per instance definition
```
Comment (dfrg) [inline fontc/src/lib.rs:1493]: Should this be “1 is 700”?

File: glyphs-reader/src/font.rs [hunk 2]
```diff
@@ -314,6 +316,51 @@ impl RawGlyph {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Hash)]
+pub struct Instance {
+    pub name: String,
+    pub active: bool,
+    pub type_: InstanceType,
```
Comment (dfrg) [inline glyphs-reader/src/font.rs:323]: nit: we commonly spell this as `ty` in Rust land. Alternatively, `kind` and `InstanceKind`.

File: glyphs-reader/src/font.rs [hunk 3]
```diff
@@ -314,6 +316,51 @@ impl RawGlyph {
...
+        }
+    }
+}
+
+#[derive(Debug, Clone, FromPlist, PartialEq, Eq, Hash)]
+struct RawInstance {
+    pub name: String,
+    pub exports: Option<i64>,
+    pub active: Option<i64>,
+    pub type_: Option<String>,
```
Comment (dfrg) [inline glyphs-reader/src/font.rs:349]: This one might be using magic in `FromPlist`?

File: glyphs-reader/src/font.rs [hunk 4]
```diff
@@ -1032,18 +1133,71 @@ fn user_to_design_from_axis_location(
...
+        return;
+    }
+    mappings.push((user, design));
+}
+
+/// * <https://github.com/googlefonts/glyphsLib/blob/6f243c1f732ea1092717918d0328f3b5303ffe56/Lib/glyphsLib/builder/axes.py#L128>
+/// * <https://github.com/googlefonts/glyphsLib/blob/6f243c1f732ea1092717918d0328f3b5303ffe56/Lib/glyphsLib/builder/axes.py#L353>
+fn add_mappings_from_instances(
+    instances: &[Instance],
+    mut mappings: BTreeMap<String, RawUserToDesignMapping>,
```
Comment (dfrg) [inline glyphs-reader/src/font.rs:1151]: Why don’t we take mappings by mutable reference here? Also, the parameter order is different from `add_mapping_if_new` which is confusing.

File: glyphs-reader/src/font.rs [hunk 5]
```diff
@@ -1032,18 +1133,71 @@ fn user_to_design_from_axis_location(
...
+            let mappings = mappings.entry(axis_name.clone()).or_default();
+            for (user, design) in inst_mapping.iter() {
+                add_mapping_if_new(mappings, *user, *design);
+            }
+        }
+    }
+    mappings
+}
+
+fn add_mappings_from_masters(
```
Comment (dfrg) [inline glyphs-reader/src/font.rs:1167]: Ditto. Maybe these mappings should be promoted to full types with methods.

File: glyphs-reader/src/font.rs [hunk 6]
```diff
@@ -1032,18 +1133,71 @@ fn user_to_design_from_axis_location(
...
+    from: &RawFont,
+    mut mappings: BTreeMap<String, RawUserToDesignMapping>,
+) -> BTreeMap<String, RawUserToDesignMapping> {
+    for master in from.font_master.iter() {
+        let Some(axes) = from.axes.as_ref() else { continue; };
+        for (axis, value) in axes.iter().zip(master.axes_values.as_ref().unwrap()) {
+            let mappings = mappings.entry(axis.name.clone()).or_default();
```
Comment (dfrg) [inline glyphs-reader/src/font.rs:1174]: Maybe use different names here for clarity?

**Review by dfrg — APPROVED** [review]
LGTM with nits. I can’t comment on glyphs format specific things but will approve to unblock until our expert returns.

File: glyphs-reader/src/font.rs [hunk 7]
```diff
@@ -557,15 +604,18 @@ fn custom_param_mut<'a>(
...
     // Per https://github.com/googlefonts/fontmake-rs/pull/42#pullrequestreview-1211619812
     // the field to use is based on the order in axes NOT the tag.
     // That is, whatever the first axis is, it's value is in the weightValue field. Long sigh.
     // Defaults per https://github.com/googlefonts/fontmake-rs/pull/42#discussion_r1044415236.
+    // v2 instances use novel field names so send back several for linear probing.
     Ok(match nth_axis {
-        0 => ("weightValue", 100_f64),
-        1 => ("widthValue", 100_f64),
-        2 => ("customValue", 0_f64),
+        0 => (&["weightValue", "interpolationWeight"], 100_f64),
```
Comment (cmyr) [inline glyphs-reader/src/font.rs:616]: the compiler should be able to infer the float type, here?

File: glyphs-reader/src/font.rs [hunk 8]
```diff
@@ -632,6 +682,34 @@ fn v2_to_v3_name(properties: &mut Vec<RawName>, v2_prop: &Option<String>, v3_nam
...
+        // Per https://github.com/googlefonts/fontmake-rs/pull/42#pullrequestreview-1211619812
+        // the field to use is based on the order in axes NOT the tag.
+        // That is, whatever the first axis is, it's value is in the weightValue field. Long sigh.
+        let (field_names, default_value) = glyphs_v2_field_name_and_default(idx)?;
+        let value = field_names
+            .iter()
+            .filter_map(|field_name| other_stuff.remove(*field_name))
+            .next()
+            .unwrap_or_else(|| Plist::Float(default_value.into()))
+            .as_f64()
```
Comment (cmyr) [inline glyphs-reader/src/font.rs:701]: Looking at this more closely, the flow is a little funny. We want to return an error if there *is* a value in the plist for this field name, but it is the wrong type (not a number?) but we want to use a default value if there is no plist value for this field name?

If that is true, I would prefer to write this as something like,

```rust
        let value = field_names
            .iter()
            .find_map(|field_name| {
                other_stuff.remove(*field_name).map(|x| {
                    x.as_f64().ok_or_else(|| {
                        Error::StructuralError(format!(
                            "Invalid value {x:?} for field {field_name}"
                        ))
                    })
                })
            })
            .unwrap_or(Ok(default_value))?;
        axis_values.push(value.into());
```

the basic idea being to generate the error where it occurs, instead of unwrapping with a known good value and validating at the end?


This is also just for illustrative purposes, no change required.

File: glyphs-reader/src/font.rs (hunk 2 above)
Comment (cmyr) [reply glyphs-reader/src/font.rs:323]: I see `ty`, `typ` and `type_` in various places, happy to standardize on one or the other 🤷

File: fontc/src/lib.rs (hunk 1 above)
Comment (rsheeter) [reply fontc/src/lib.rs:1493]: Yes, oops

File: glyphs-reader/src/font.rs [hunk 9]
```diff
@@ -1032,18 +1133,71 @@ fn user_to_design_from_axis_location(
...
+fn add_mapping_if_new(
+    mappings: &mut RawUserToDesignMapping,
+    user: OrderedFloat<f32>,
+    design: OrderedFloat<f32>,
+) {
+    if mappings.iter().any(|(u, d)| *u == user || *d == design) {
+        return;
+    }
+    mappings.push((user, design));
+}
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply glyphs-reader/src/font.rs:1145]: In re-reading I think I agree ... that deserves to be a "real" type now.

File: glyphs-reader/src/font.rs [hunk 10]
```diff
@@ -929,7 +1019,18 @@ fn default_master_idx(raw_font: &RawFont) -> usize {
...
+        .unwrap_or_else(|| {
+            raw_font
+                .font_master
+                .iter()
+                .position(|m| match m.other_stuff.get("name") {
+                    Some(Plist::String(name)) => name == "Regular",
+                    _ => false,
+                })
+                .unwrap_or_default()
+        })
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply glyphs-reader/src/font.rs:1033]: Oh nice, I'm not used to if in matches yet so I underuse it.

File: glyphs-reader/src/font.rs [hunk 11]
```diff
@@ -1172,6 +1326,110 @@ fn raw_feature_to_feature(feature: RawFeature) -> Result<FeatureSnippet, Error>
     Ok(FeatureSnippet(code))
 }
 
+/// <https://github.com/googlefonts/glyphsLib/blob/6f243c1f732ea1092717918d0328f3b5303ffe56/Lib/glyphsLib/classes.py#L220-L249>
+fn lookup_class_value(axis_tag: &str, user_class: &Option<&String>) -> Option<f32> {
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply glyphs-reader/src/font.rs:1326]: I think I confused myself and it should be &str. Fixed.

File: glyphs-reader/src/font.rs [hunk 12]
```diff
@@ -1172,6 +1326,110 @@ fn raw_feature_to_feature(feature: RawFeature) -> Result<FeatureSnippet, Error>
...
+        }
+    }
+}
+
+fn add_mapping_if_present(
+    axis_mappings: &mut BTreeMap<String, RawUserToDesignMapping>,
+    axes: &[Axis],
+    axis_tag: &str,
+    axes_values: &Option<Vec<OrderedFloat<f64>>>,
+    value: &Option<&Plist>,
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply glyphs-reader/src/font.rs:1373]: Updated to use `Option<&whatever>`

File: glyphs-reader/src/font.rs (hunk 7 above)
Comment (rsheeter) [reply glyphs-reader/src/font.rs:616]: yup, ty

File: glyphs-reader/src/font.rs [hunk 13]
```diff
@@ -632,6 +682,34 @@ fn v2_to_v3_name(properties: &mut Vec<RawName>, v2_prop: &Option<String>, v3_nam
...
+    let mut axis_values = Vec::new();
+    for idx in 0..axes.len() {
+        // Per https://github.com/googlefonts/fontmake-rs/pull/42#pullrequestreview-1211619812
+        // the field to use is based on the order in axes NOT the tag.
+        // That is, whatever the first axis is, it's value is in the weightValue field. Long sigh.
+        let (field_names, default_value) = glyphs_v2_field_name_and_default(idx)?;
+        let value = field_names
+            .iter()
+            .filter_map(|field_name| other_stuff.remove(*field_name))
+            .next()
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply glyphs-reader/src/font.rs:699]: TIL! Yes :)

File: glyphs-reader/src/font.rs [hunk 14]
```diff
@@ -314,6 +316,51 @@ impl RawGlyph {
...
+
+#[derive(Debug, Clone, FromPlist, PartialEq, Eq, Hash)]
+struct RawInstance {
+    pub name: String,
+    pub exports: Option<i64>,
+    pub active: Option<i64>,
+    pub type_: Option<String>,
+    pub axes_values: Option<Vec<OrderedFloat<f64>>>,
+
+    #[rest]
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply glyphs-reader/src/font.rs:352]: TIL. I think I will fix that in it' sown PR.

File: glyphs-reader/src/font.rs (hunk 8 above)
Comment (rsheeter) [reply glyphs-reader/src/font.rs:701]: In re-reading I think I will change to issue a warning if it's a weird value

File: glyphs-reader/src/font.rs (hunk 3 above)
Comment (rsheeter) [reply glyphs-reader/src/font.rs:349]: This one I believe we need to name so to make FromPlist work, will add comment

File: glyphs-reader/src/font.rs (hunk 2 above)
Comment (rsheeter) [reply glyphs-reader/src/font.rs:323]: These names have to match the glyphs data and work with FromPlist, will add a comment

File: glyphs-reader/src/font.rs (hunk 4 above)
Comment (rsheeter) [reply glyphs-reader/src/font.rs:1147]: I wanted to be able to write `add_mappings(blah)` in return position. In the light of a new day i don't think it's a win, will change to mut ref.

Comment (rsheeter) [reply glyphs-reader/src/font.rs:1147]: ...actually to it's own type

File: glyphs-reader/src/font.rs [hunk 15]
```diff
@@ -314,6 +316,51 @@ impl RawGlyph {
...
+    }
+}
+
+#[derive(Debug, Clone, FromPlist, PartialEq, Eq, Hash)]
+struct RawInstance {
+    pub name: String,
+    pub exports: Option<i64>,
+    pub active: Option<i64>,
+    pub type_: Option<String>,
+    pub axes_values: Option<Vec<OrderedFloat<f64>>>,
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply glyphs-reader/src/font.rs:350]: Correct. It would be really nice if FromPlist could deserialize to an empty vec but I don't want to tackle that in this PR. 

Filed as https://github.com/googlefonts/fontmake-rs/issues/257

File: glyphs-reader/src/font.rs (hunk 6 above)
Comment (rsheeter) [reply glyphs-reader/src/font.rs:1170]: Became nicer as a result of adding types

File: glyphs-reader/src/font.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (rsheeter) [reply glyphs-reader/src/font.rs:358]: https://github.com/googlefonts/fontmake-rs/pull/259

File: fontc/src/lib.rs [hunk 16]
```diff
@@ -706,6 +706,7 @@ mod tests {
         fs::{self, File},
         io::Read,
         path::{Path, PathBuf},
+        str::FromStr,
```
Comment (anthrotype) [inline fontc/src/lib.rs:709]: noob question: why is this needed?

Comment (dfrg) [reply fontc/src/lib.rs:709]: A couple of the tests use `Tag::from_str` and the `FromStr` trait needs to be in scope for that method to be available.

Comment (anthrotype) [reply fontc/src/lib.rs:709]: Interesting. AFAIU, the Tag struct implements `FromStr` trait, which comes from `std::str::FromStr`; what it's not immediately clear to me is why the users of that method need to import that trait, and it is not enough to simply import Tag.

**Review by anthrotype — COMMENTED** [review]
it must have been not fun to digest the messy logic from glyphsLib... LGTM!

File: fontc/src/lib.rs (hunk 16 above)
Comment (dfrg) [reply fontc/src/lib.rs:709]: Just the rules of the language. The trait itself must be in scope in order to call its methods. You can implement multiple traits with the same method name(s) for a single type and this is one way of disambiguating those cases.

This is sometimes muddled because Rust has a [prelude](https://doc.rust-lang.org/std/prelude/index.html) that automatically imports common types and traits to all modules. So you’ll not see traits imported explicitly if they appear there.

File: glyphs-reader/src/font.rs [hunk 17]
```diff
@@ -929,7 +1020,14 @@ fn default_master_idx(raw_font: &RawFont) -> usize {
                 .map(|(idx, _)| idx)
                 .unwrap_or(0)
         })
-        .unwrap_or(0)
+        // TODO: implement searching for "a base style shared between all masters" as FontTools does
+        // Still nothing? - just look for one called Regular
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply glyphs-reader/src/font.rs:1024]: For Oswald fontmake falls back to Regular after failing at base style matching so this is incrementally closer. I need to find - or create - a source we build wrong due to not having the full find a shared base style. Will file an issue.

File: glyphs-reader/src/font.rs [hunk 18]
```diff
@@ -929,7 +1020,14 @@ fn default_master_idx(raw_font: &RawFont) -> usize {
                 .map(|(idx, _)| idx)
                 .unwrap_or(0)
         })
-        .unwrap_or(0)
+        // TODO: implement searching for "a base style shared between all masters" as FontTools does
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply glyphs-reader/src/font.rs:1023]: Fixed by https://github.com/googlefonts/fontmake-rs/pull/265

File: glyphs-reader/src/font.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (rsheeter) [reply glyphs-reader/src/font.rs:1024]: https://github.com/googlefonts/fontmake-rs/issues/264
